rootston/cursor.c

```diff
@@ -165,14 +166,20 @@ static void roots_cursor_press_button(struct roots_cursor *cursor,
 		return;
 	}
 
-	uint32_t serial =
-		wlr_seat_pointer_notify_button(seat->seat, time, button, state);
+	uint32_t serial;
+	if (is_touch) {
+		serial = wl_display_get_serial(desktop->server->wl_display);
```

Comment:
I especially don't like this `wl_display_get_serial` because:

1. It's in rootston, meaning it's not internal stuff
2. We really don't understand where the `serial` is coming from

Is there a way to do something similar to pointers?

Reply:
-1, this is how you do serials

Reply:
>this is how you do serials

You do serials by incrementing them in functions (and let the user guess which function increments the serial) and then read them in a completely separate function (which would fail if not called by a function that in fact incremented the serial), hoping for it not to have been incremented by another function?

Reply:
In wayland you do serials with wl_display_get_serial.

Reply:
I agree that we should be getting the serial directly from the down event. If you look at the touch down handler in rootston, you will see that we are guaranteed to get the right serial because the down notify happens right before the call to press the button and won't run if no down event was actually sent.

I think it's good enough for now. This will get refactored into something nicer when we flesh out all of the compositor touch actions in a future PR.

Reply:
It really, really doesn't matter. It just has to be a unique number and wl_display_get_serial does that.

Reply:
Wait, `input_events` doesn't seem used anymore. Should we remove it?

>It really, really doesn't matter. It just has to be a unique number and wl_display_get_serial does that.

It should be the serial of the touch event, not just a unique number. Note that we're using `wl_display_get_serial`, not `wl_display_next_serial`.

Reply:
In this particular case, the serial is used to validate a moveresize request from the shells. The risk is that if you don't get the right serial from the down event, then whatever serial you actually got instead is used for validation which can lead to false positives/negatives. The code how it is will always be correct, but @emersion 's concerns that this code is not robust to changes is valid and should be considered in the upcoming refactor.

Reply:
Oh yeah that sucks, I was mistaking it for next_serial